examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -850,147 +1014,318 @@ void CassieRbtStateEstimator::contactEstimation(
+      }
+    }
+
+    // Estimated floating base state (pelvis)
+    VectorXd fb_state_est(13);
+    Vector3d r_imu_to_pelvis_global =
+        filter.getState().getRotation() * (-imu_pos_);
+    fb_state_est.head(3) =
+        filter.getState().getPosition() + r_imu_to_pelvis_global;
+    Quaterniond q(filter.getState().getRotation());
+    q.normalize();
+    fb_state_est[3] = q.w();
+    fb_state_est.segment<3>(4) = q.vec();
+    fb_state_est.segment<3>(7) = imu_measurement.head(3);
+    Vector3d omega_global =
+        filter.getState().getRotation() * imu_measurement.head(3);
+    fb_state_est.tail(3) = filter.getState().getVelocity() +
+        omega_global.cross(r_imu_to_pelvis_global);
+
+    // Estimated robot output
+    OutputVector<double> filtered_output(tree_.get_num_positions(),
+                                   tree_.get_num_velocities(),
+                                   tree_.get_num_actuators());
+    AssignImuValueToOutputVector(cassie_out, &filtered_output);
+    AssignActuationFeedbackToOutputVector(cassie_out, &filtered_output);
+    AssignNonFloatingBaseStateToOutputVector(cassie_out, &filtered_output);
+    AssignFloatingBaseStateToOutputVector(fb_state_est, &filtered_output);
+
+    // Step 3 - Estimate which foot/feet are in contact with the ground
     // Estimate feet contacts
     int left_contact = 0;
     int right_contact = 0;
-    contactEstimation(output, dt,
-        discrete_state, &left_contact, &right_contact);
-
-    // Step 4 - EKF (measurement step)
+    if (test_with_ground_truth_state_){
+      UpdateContactEstimationCosts(output_gt, dt,
+          &(state->get_mutable_discrete_state()));
+      EstimateContactForEkf(output_gt, &left_contact, &right_contact);
+    } else {
+      UpdateContactEstimationCosts(filtered_output, dt,
+          &(state->get_mutable_discrete_state()));
+      EstimateContactForEkf(filtered_output, &left_contact, &right_contact);
+    }
 
+    std::vector<std::pair<int, bool>> contacts;
+    contacts.push_back(std::pair<int, bool>(0, left_contact));
+    contacts.push_back(std::pair<int, bool>(1, right_contact));
+    filter.setContacts(contacts);
 
-    // Step 5 - Assign values to states
-    // Below is how you should assign the state at the end of this Update
-    // discrete_state->get_mutable_vector(ekf_X_idx_).get_mutable_value() = ...;
-    // discrete_state->get_mutable_vector(time_idx_).get_mutable_value() = ...;
+    // Step 4 - EKF (measurement step)
+    KinematicsCache<double> cache = tree_.doKinematics(
+        filtered_output.GetPositions(), filtered_output.GetVelocities(), true);
+    Vector3d d, p;
+    // rotation part of pose and covariance is unused in EKF
+    Eigen::Matrix4d pose = Eigen::Matrix4d::Identity();
+    Eigen::Matrix<double, 6, 6> covariance = MatrixXd::Identity(6, 6);
+    Eigen::Matrix<double, 22, 22> cov_w =
+        0.000289 * MatrixXd::Identity(22, 22); // nosie of joints measurement
```

Comment:
Should this constant be set with the rest of the constants, rather than inline here?

Reply:
Yes, moved to constructor.